cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/ProgramScheduleStoreDataset.java

```diff
@@ -0,0 +1,232 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.internal.app.runtime.schedule.store;
+
+import co.cask.cdap.api.common.Bytes;
+import co.cask.cdap.api.dataset.DatasetSpecification;
+import co.cask.cdap.api.dataset.lib.AbstractDataset;
+import co.cask.cdap.api.dataset.lib.IndexedTable;
+import co.cask.cdap.api.dataset.module.EmbeddedDataset;
+import co.cask.cdap.api.dataset.table.Delete;
+import co.cask.cdap.api.dataset.table.Get;
+import co.cask.cdap.api.dataset.table.Put;
+import co.cask.cdap.api.dataset.table.Row;
+import co.cask.cdap.api.dataset.table.Scan;
+import co.cask.cdap.api.dataset.table.Scanner;
+import co.cask.cdap.common.AlreadyExistsException;
+import co.cask.cdap.internal.app.runtime.schedule.ProgramSchedule;
+import co.cask.cdap.internal.app.runtime.schedule.trigger.PartitionTrigger;
+import co.cask.cdap.internal.app.runtime.schedule.trigger.Trigger;
+import co.cask.cdap.internal.app.runtime.schedule.trigger.TriggerJsonDeserializer;
+import co.cask.cdap.proto.id.ApplicationId;
+import co.cask.cdap.proto.id.ProgramId;
+import co.cask.cdap.proto.id.ScheduleId;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import javax.annotation.Nullable;
+
+/**
+ * Dataset that stores and oindexes proram schedules, so that they can be looked by their trigger keys.
+ */
+public class ProgramScheduleStoreDataset extends AbstractDataset {
+
+  private static final String SCHEDULE_COLUMN = "sch";
+  private static final byte[] SCHEDULE_COLUMN_BYTES = Bytes.toBytes(SCHEDULE_COLUMN);
+  private static final String TRIGGER_KEY_COLUMN = "tk"; // trigger key
+  private static final byte[] TRIGGER_KEY_COLUMN_BYTES = Bytes.toBytes(TRIGGER_KEY_COLUMN);
+
+  // package visible for the dataset definition
+  static final String EMBEDDED_TABLE_NAME = "it"; // indexed table
+  static final String INDEX_COLUMNS = TRIGGER_KEY_COLUMN; // trigger key
+
+  private static final Gson GSON = new GsonBuilder().registerTypeAdapter(Trigger.class, new TriggerJsonDeserializer()).create();
+
+  private final IndexedTable store;
+
+  public ProgramScheduleStoreDataset(DatasetSpecification spec,
+                                     @EmbeddedDataset(EMBEDDED_TABLE_NAME) IndexedTable store) {
+    super(spec.getName(), store);
+    this.store = store;
+  }
+
+  /**
+   * Add one or more schedules to the store.
+   *
+   * @param schedules the schedules to add
+   *
+   * @throws AlreadyExistsException if one of the schedules already exists
+   */
+  public void addSchedules(ProgramSchedule... schedules) throws AlreadyExistsException {
```

Comment:
Might be easier to call if having a overload that takes `Collection<? extends ProgramSchedule>` as well (from app creation, most likely it'll have a collection, while from REST API, most likely it'll be single item).